src/wallet/rpcdisclosure.cpp

```diff
@@ -0,0 +1,296 @@
+        throw runtime_error(
+            "z_validatepaymentdisclosure \"paymentdisclosure\"\n"
+            "\nValidates a payment disclosure.\n"
+            "\nEXPERIMENTAL FEATURE\n"
+            + strPaymentDisclosureDisabledMsg +
+            "\nArguments:\n"
+            "1. \"paymentdisclosure\"     (string, required) Hex data string\n"
+            "\nExamples:\n"
+            + HelpExampleCli("z_validatepaymentdisclosure", "\"hexblob\"")
+            + HelpExampleRpc("z_validatepaymentdisclosure", "\"hexblob\"")
+        );
+
+    if (!fEnablePaymentDisclosure) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Error: payment disclosure is disabled.");
+    }
+
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+
+    EnsureWalletIsUnlocked();
+
+    string hexInput = params[0].get_str();
+    if (!IsHex(hexInput))
+    {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, expected payment disclosure data in hexadecimal format.");
+    }
+
+    // Unserialize the payment disclosure data into an object
+    PaymentDisclosure pd;
+    CDataStream ss(ParseHex(hexInput), SER_NETWORK, PROTOCOL_VERSION);
+    try {
+        ss >> pd;
+        // too much data is ignored, but if not enough data, exception of type ios_base::failure is thrown,
+        // CBaseDataStream::read(): end of data: iostream error
+    } catch (const std::exception &e) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, payment disclosure data is malformed.");        
+    }
+
+    uint256 hash = pd.payload.txid;
+    CTransaction tx;
+    uint256 hashBlock;
+    // Check if we have seen the transaction
+    if (!GetTransaction(hash, tx, hashBlock, true)) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "No information available about transaction");
+    }
+
+    // Check if the transaction has been confirmed
+    if (hashBlock.IsNull()) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Transaction has not been confirmed yet");
+    }
+
+    // Check if shielded tx
+    if (tx.vjoinsplit.empty()) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Transaction is not a shielded transaction");        
+    }
+
+    UniValue errs(UniValue::VARR);
+    UniValue o(UniValue::VOBJ);
+    o.push_back(Pair("txid", pd.payload.txid.ToString()));
+
+    // Check js_index
+    if (pd.payload.js >= tx.vjoinsplit.size()) {
+        errs.push_back("Payment disclosure refers to an invalid joinsplit index");
+    }
+    o.push_back(Pair("jsIndex", pd.payload.js));
+
+    if (pd.payload.n < 0 || pd.payload.n >= ZC_NUM_JS_OUTPUTS) {
+        errs.push_back("Payment disclosure refers to an invalid output index");
+    }
+    o.push_back(Pair("outputIndex", pd.payload.n));
+
+    if (pd.payload.version != PAYMENT_DISCLOSURE_VERSION_EXPERIMENTAL) {
+        errs.push_back("Payment disclosure refers to an unknown version");
+    }
+    o.push_back(Pair("version", pd.payload.version));
+
+    o.push_back(Pair("onetimeSecretKey", pd.payload.esk.ToString()));
```

Comment:
I'd prefer "onetimePrivateKey".

Reply:
We discussed earlier that esk was out, because we call the public portion in json output 'onetimePubKey'.  So I've changed this to 'onetimePrivKey' now.  I think this matches the language used in the rest of RPC interface e.g. 'importprivkey' where Public and Private are shortened.